extensions/yaml-configuration/runtime/src/main/java/io/quarkus/yaml/configuration/runtime/YamlConfigSource.java

```diff
@@ -0,0 +1,125 @@
+
+public abstract class YamlConfigSource implements ConfigSource, Serializable {
+
+    private static final long serialVersionUID = -875865410658150582L;
+
+    private static final String CONFIG_ORDINAL_KEY = "config_ordinal";
+    private static final String CONFIG_ORDINAL_DEFAULT_VALUE = "100";
+    private static final String PROPERTY_KEY_SEPARATOR = ".";
+    private static final String COMMA = ",";
+    private static final String ESCAPED_COMMA = "\\\\,";
+
+    private final Map<String, String> properties;
+    private final String source;
+    private final int ordinal;
+
+    protected YamlConfigSource(String source, int ordinal) {
+        this.properties = buildConfigPropertiesFromYaml();
+        this.source = source;
+        if (properties.containsKey(CONFIG_ORDINAL_KEY)) {
+            this.ordinal = Integer.valueOf(properties.getOrDefault(CONFIG_ORDINAL_KEY, CONFIG_ORDINAL_DEFAULT_VALUE));
+        } else {
+            this.ordinal = ordinal;
+        }
+    }
+
+    protected abstract InputStream openYamlFileStream();
+
+    private Map<String, String> buildConfigPropertiesFromYaml() {
+        try (InputStream is = openYamlFileStream()) {
+            if (is == null) {
+                return Collections.emptyMap();
+            }
+            @SuppressWarnings("unchecked")
+            Map<String, Object> yamlContent = new Yaml().loadAs(is, HashMap.class);
+            Map<String, String> configProperties = new HashMap<>();
+            for (YamlConfigProperty prop : convertYamlMapToProperties("", yamlContent)) { // The root `baseKey` is empty.
+                configProperties.put(prop.key, prop.value);
+            }
+            return configProperties;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private List<YamlConfigProperty> convertYamlMapToProperties(String baseKey, Object object) {
+        if (object instanceof Map) {
+            List<YamlConfigProperty> properties = new ArrayList<>();
+            for (Entry<?, ?> entry : ((Map<?, ?>) object).entrySet()) {
+                String entryKey = baseKey + (baseKey.isEmpty() ? "" : PROPERTY_KEY_SEPARATOR) + entry.getKey();
+                properties.addAll(convertYamlMapToProperties(entryKey, entry.getValue()));
+            }
+            return properties;
+        } else {
+            String value;
+            if (object instanceof List) {
+                value = stringifyList(object);
+            } else {
+                value = object.toString();
+            }
+            return Collections.singletonList(new YamlConfigProperty(baseKey, value));
+        }
+    }
+
+    private String stringifyList(Object list) {
+        List<String> stringifiedElements = new ArrayList<>();
+        for (Object elem : (List<?>) list) {
+            String stringifiedElem = String.valueOf(elem);
+            if (stringifiedElem.contains(COMMA)) {
+                stringifiedElem.replace(COMMA, ESCAPED_COMMA);
```

Comment:
`Pattern` compilation could improve the performances here with JDK 8, but the implementation of `String.replace()` was changed in both JDK 9 and 13 and made the `Pattern` compilation useless. I chose to favor the most recent JDKs.

Reply:
Except that Strings are immutable and you forgot to assign the return value to `stringifiedElem` :)

Reply:
Thanks for pointing this out! The return value assignment was there at first, I must have missed a copy/paste at some point :) I really need to add that integration test.

Reply:
Fixed, thanks!